### PR TITLE
Add plots to illustrate what the optimizer does

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: docs/source/dataframe-optimizer.rst [hunk 1]
```diff
@@ -38,13 +38,25 @@ The optimizations entail the following steps (this list is not complete):
   ``df.groupby(...).apply(...)`` after a merge operation will not shuffle the data
   again if the groupby happens on the merge columns.
 
+  .. figure:: images/optimizer/avoiding-shuffles.svg
+    :align: center
+
+    Performing two subsequent Joins/Merges on the same join-column(s) will avoid shuffling the
+    data again. The optimizer identifies that the partitioning of the DataFrame is already as
+    expected and thus simplifies the operation to a single Shuffle and a trivial merge operation.
```
Comment (hendrikmakait) [inline docs/source/dataframe-optimizer.rst:46]: nit: The term "trivial" merge seems very vague to me, maybe "pre-partitioned" merge highlights what's special/trivial about it?
```suggestion
    expected and thus simplifies the operation to a single Shuffle and a pre-partitioned merge operation.
```

File: docs/source/images/optimizer/automatic-repartitioning.svg
Comment (hendrikmakait) [inline docs/source/images/optimizer/automatic-repartitioning.svg:1]: nit: If we're going for a hand-drawn look, I'd recommend using a matching font (and illustration for the files, if easy).

File: docs/source/dataframe-optimizer.rst (hunk 1 above)
Comment (hendrikmakait) [inline docs/source/dataframe-optimizer.rst:46]: How does this work again? Do we only detect this for subsequent merges or all cases where the data is correctly pre-partitioned?

Comment (phofl) [reply docs/source/dataframe-optimizer.rst]: All cases where we partitioned at some point in the past, e.g. 

result = df.merge(df2)
result = result.dropna()
result = result.merge(df3)

would work as well.

File: docs/source/images/optimizer/automatic-repartitioning.svg
Comment (phofl) [reply docs/source/images/optimizer/automatic-repartitioning.svg:1]: I don't see a difference here, can you elabortate?

File: docs/source/dataframe-optimizer.rst [hunk 2]
```diff
@@ -38,13 +38,25 @@ The optimizations entail the following steps (this list is not complete):
   ``df.groupby(...).apply(...)`` after a merge operation will not shuffle the data
   again if the groupby happens on the merge columns.
 
+  .. figure:: images/optimizer/avoiding-shuffles.svg
+    :align: center
+
+    Performing two subsequent Joins/Merges on the same join-column(s) will avoid shuffling the
```
Comment (hendrikmakait) [inline docs/source/dataframe-optimizer.rst:44]: nit: IMO this makes it clearer that this is still an example and this optimization is not limited to these cases. YMMV, feel free to ignore.

```suggestion
    Similarly, performing two subsequent Joins/Merges on the same join-column(s) will avoid shuffling the
```

File: docs/source/images/optimizer/automatic-repartitioning.svg
Comment (hendrikmakait) [reply docs/source/images/optimizer/automatic-repartitioning.svg:1]: I don't have access to the source files, so I can't suggest an actual alternative plot. :(

My style suggestion is still a nit though, so feel free to ignore.

File: docs/source/images/optimizer/avoiding-shuffles.svg
Comment (hendrikmakait) [inline docs/source/images/optimizer/avoiding-shuffles.svg:1]: nit: Maybe replace "Trivial Join" with "Pre-Partitioned Join"?

Comment (hendrikmakait) [reply docs/source/images/optimizer/avoiding-shuffles.svg:1]: In that case I suggest reverting https://github.com/dask/dask/pull/11072/commits/cffe539944d6cf4b1b3061af17720fddcee8d988 to stay consistent.

Comment (phofl) [reply docs/source/images/optimizer/avoiding-shuffles.svg:1]: done

Comment (mrocklin) [reply docs/source/images/optimizer/avoiding-shuffles.svg:1]: +1 on Trivial over Pre-Partitioned.  My guess is that most readers won't understand what PrePartitioned means.

File: docs/source/images/optimizer/automatic-repartitioning.svg
Comment (phofl) [reply docs/source/images/optimizer/automatic-repartitioning.svg:1]: done

**Pull request merged**